### ISSUE TITLE
How to make every class name to be unique?

### ISSUE THREAD

**Issue by Profesor08** [cristianbote/goober#397]
I have a situation where I want to apply style to element on some case. But I can't target desired child element, becaouse a lot of child elements has the same class name. `.go11` is child of `.go11` wtf?
![image](https://user-images.githubusercontent.com/8218016/144035146-38129a30-98be-481a-8d2b-4edca73066e9.png)

**Comment by B-Teague** [comment]
@Profesor08 See my reply for https://github.com/cristianbote/goober/issues/387
The current workaround is to refactor your code to move different hover values into the children instead of the parent.
Ideally, you don't want two components to have the exact same css.  Once their css is different, they can become their own components.  Example: HoverLinkRed and HoverLinkBlue

**Comment by Profesor08** [comment]
> The current workaround is to refactor your code to move different hover values into the children instead of the parent.

This is required behavior. Some elements must change their styles on hovering their container. This is very common case, where you need to change something inside depending on hover. Showing/hiding some buttons, replacing one lement with onother. Some times they has different styles, but some times thei have same styles, or just has some common wrapper.

Making them to extend some base styled do nothing, bug remains. Because hover targets their common className.

The one solution to make all this to work correctly is to assigin unique classname to element on `styled()` function call. Because every time function `styled()` is called, it creates an new component. And this new component is not the same as some component previously declared. 

One more example with different elements. You can't just extends one from another. And they both must change their styles depending on parent `:hover`. There is one more bug. Styles priority are disrespected. purple color for link must have highter priority and remain purple.
```
const Link = styled("a")``;

const Button = styled("div")``;

const ControlWrapper = styled("div")`
  width: 100px;

  ${Link} {
    color: purple;
  }
`;

const LinkControl = styled(({ children }) => {
  return (
    <ControlWrapper>
      <Link>{children}</Link>
    </ControlWrapper>
  );
})``;

const Parent = styled("div")`
  &:hover {
    ${Link} {
      color: red;
    }
    ${Button} {
      color: blue;
    }
    ${ControlWrapper} {
      color: green;
    }
  }
`;

function App() {
  return (
    <Parent>
      <Link>ChildA</Link>
      <Button>ChildB</Button>
      <LinkControl>LinkControl</LinkControl>
    </Parent>
  );
}
```

**Comment by B-Teague** [comment]
@cristianbote There is definitely different behavior here.  Just so we can compare apples and oranges, I cloned @Profesor08  example using styled-components instead of goober to compare functionality:

https://codesandbox.io/s/react-goober-forked-ynd84?file=/src/index.js

You can comment out the setup line and goober and enable styled-components to see how the color changes differently on hover.

@Profesor08 The LinkControl changes to red using styled-components.  Were you expecting it to turn red or stay purple?

**Comment by B-Teague** [comment]
So it looks like StyledComponents doesn't cache it's CSS?  Since every styled component has a unique class, this allows you to use the same CSS across multiple styled components without conflict.

**Comment by B-Teague** [comment]
@Profesor08 A hacky workaround that I tested is to put in some dummy css to force the template strings to be different between Link and Button.  This will generate unique class names which should address your problem.
@cristianbote I think giving the developer the ability to force a unique class name when the css is the same might be a necessary feature to address this problem.  Your thoughts?

I added this feature to morpho which fixed this exact problem
* prefix - This option is used to specify the global custom prefix for generated css class names. Can be overridden by each css function call css({}, "customPrefix"). Default value is "morpho".

```javascript
var key = prefix + rules.join("")
return cache[key] || (cache[key] = createStyle(rules, cssType, prefix))
```

**Comment by cristianbote** [comment]
Heyo!

So the `go11` className comes from an empty style rule, which makes sense. I believe the community wanted this behaviour before as well so this is why this exists:
https://github.com/cristianbote/goober/blob/168bbc55424ae3db95aca09c4fd02716efb98b78/src/styled.js#L31 would this be something that could help out?

**Comment by Profesor08** [comment]
@cristianbote 
> So the go11 className comes from an empty style rule, which makes sense.
It is ok, if you manage only style rules, and you can continue doing it by this way. But this will not work for components. Every component is designed to be unique, and reusable. They all can have common styles, common classes, but every one must have his own unique class name. And all relations between components must use only their unique class names.

**Comment by B-Teague** [comment]
@cristianbote This makes total sense when you think about it like instances of an object class.  Each styled component is its own class: Link, Button, Tab, etc.
You don't want Link and Button sharing the same css className, each instance of a Link should only use the Link css className.  Each instance of the Button should only use the Button css className.  

We need a way to cache the css "per styled call"

**Comment by Profesor08** [comment]
@B-Teague 
> Do we need a way to cache the css "per styled call"?

Yes. If some prop will change, and after that will change back, we don't need to generate every time one new class name.

> What would happen if we just remove caching of css classNames?

It will generate new class name on every render.

**Comment by B-Teague** [comment]
@Profesor08 Looks like it works as intended now assuming @cristianbote doesn't have any concerns with my PR.
https://codesandbox.io/s/react-goober-forked-3vnqd

**Comment by B-Teague** [comment]
@cristianbote I opened PR #401 which reduces the bundle size and also fixes this issue.

**Comment by cristianbote** [comment]
So, what is expected here? I'm leaning towards that this is an intended behaviour and outside of the scope of goober as I believe this is not doable from a runtime POV. Thoughts?

**Comment by B-Teague** [comment]
@cristianbote Agreed, per my last comment on the closed PR:
"I think we can state the library is working as designed unless you want to implement a way to generate unique class names for each time styled is called, but still preserve hydration for server side rendering."  

Hydration would break if the order of Button1 and Button3 are switched if we used an internal counter.  The only way to support this at runtime is evaluating the source code at runtime and using the name of the variable as part of the hashing algorithm which would just add bloat to the library with basically zero advantages.

const Button1 = styled("div")``;
const Button2 = styled("div")`color: red`;
const Button3 = styled("div")``;
